include/circt/Dialect/Calyx/CalyxStructure.td

```diff
@@ -9,3 +9,124 @@
 // This describes the Calyx structures.
 //
 //===----------------------------------------------------------------------===//
+
+def ComponentOp : CalyxOp<"component", [
+    Symbol,
+    FunctionLike,
+    IsolatedFromAbove,
+    // TODO(Calyx): Need to eventually expose the output ports.
+    NoTerminator,
+    SingleBlock
+  ]> {
+  let summary = "Calyx Component";
+  let description = [{
+    The "calyx.component" operation represents an overall
+    Calyx component containing:
+    (1) In- and output port definitions
+        that define the interface.
+    (2) The cells, wires, and control schedule.
+    ```mlir
+      component MyComponent(in1: 32) -> (out1: 32) {
+        cells { ... }
+        wires { ... }
+        control { ... }
+      }
+    ```
+  }];
+
+  // TODO(Calyx): Allow explicit port naming?
+  let arguments = (ins);
+  let results = (outs);
+
+  let regions = (region SizedRegion<1>: $body);
+
+  let extraClassDeclaration = [{
+    // Necessary to avoid name clashing with `front`.
+    using FunctionLike::front;
+
+    private:
+      // This trait needs access to the hooks defined below.
+      friend class OpTrait::FunctionLike<ComponentOp>;
+
+      /// Hooks for the input/output type enumeration in FunctionLike.
+      unsigned getNumFuncArguments() { return getType().getNumInputs(); }
+      unsigned getNumFuncResults() { return getType().getNumResults(); }
+    public:
+      /// Returns the body of a Calyx component.
+      Block *getBody() { return &getOperation()->getRegion(0).front(); }
+  }];
+
+  let printer = "return ::print$cppClass(p, *this);";
```

Comment:
Not required, but you may be able to simplify some of the parsing and printing using the declarative assembly format. You could potentially use a custom directive for the ports, and otherwise use something like:

```let assemblyFormat = "attr-dict $sym_name custom<PortDefList>() $body";```

Then you should be able to drop everything but the `parsePortDefList` and `printPortDefList`. Note that you would also have to declare `SymbolNameAttr:$sym_name` explicitly in the `arguments`. If you chose to have an explicit attribute for port names, it could go into `arguments` too, and be passed to the `custom` directive.

Reply:
Very cool, I wasn't aware of this. Thanks!

Reply:
Taking some time to look at other `custom<>` formats, I'm going to stick to using a custom printer and parser for `componentOp` for now. I also foresee a few more additions to this op (e.g. direction/width information, similar to what `firrtl` and `hw` do with `getModulePortInfo`).